### PR TITLE
testing(dot/core): rewrite dot/core service tests (part 1)

### PR DESCRIPTION
**Pull request by jimjbrettj**
## Changes

- Test half of dot/core/service.go (second half will be in separate PR). This PR includes the functions: StorageRoot, HandleBlockProduced, handleBlock, handleCodeSubstitution, handleBlockAsync, and maintainTransactionPool
- Uncommented necessary code and removed TODO from maintainTransactionPool

## Tests

```
go test ./dot/core -v -coverprofile coverage.out
go tool cover -func coverage.out
```

## Issues

- #2184

## Primary Reviewer

- @noot

### CODE REVIEW

File: dot/core/service_test.go [hunk 1]
```diff
@@ -0,0 +1,616 @@
...
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			s := tt.service
+			s.handleBlocksAsync()
+		})
+	}
+}
+
+func TestCleanup(t *testing.T) {
```
Comment (timwu20) [inline dot/core/service_test.go:613]: what is this for?

Comment (jimjbrettj) [reply dot/core/service_test.go:613]: The tests need to generate and remove wasm files, but cant have a TestMain function to handle it since it already exists in this packages integration tests. My attempt of a work around was to have a method to generate the wasm then this TestCleanup function to remove the generated wasm.

Comment (noot) [reply dot/core/service_test.go:613]: maybe just name it `cleanup` then? since it's not really a test

**Review by noot — APPROVED** [review]

File: dot/core/service_test.go (hunk 1 above)
Comment (jimjbrettj) [reply dot/core/service_test.go:613]: Will it run automatically still? If not should I just call it at the end of the last test?

Comment (jimjbrettj) [reply dot/core/service_test.go:613]: I can make this change, but then if any tests are added whoever adds them will have to remember to move the call to cleanup() to the last test run. @noot is this ideal or should I leave as is?

Comment (noot) [reply dot/core/service_test.go:613]: I think it's better to call it at the end of every test that uses it in a t.Cleanup(), just to make it clearer when it's run

Comment (jimjbrettj) [reply dot/core/service_test.go:613]: So for every test I should generate and take down the wasm? Sounds cleaner but would definitely add a decent amount of time to the tests

File: dot/core/service_test.go [hunk 2]
```diff
@@ -0,0 +1,616 @@
...
+			err = s.handleCodeSubstitution(tt.args.hash, tt.args.state)
+			assert.ErrorIs(t, err, tt.expErr)
+			if tt.expErr != nil {
+				assert.EqualError(t, err, tt.expErrMsg)
+			}
+		})
+	}
+}
+
+func TestService_handleBlock(t *testing.T) {
```
Comment (qdm12) [inline dot/core/service_test.go:188]: Improve readability with 2 ways:

### Mocks configuration changing a lot from a test case to another

⚠️ this is usually a code smell showing the production function being tested is too long/complex.

BUT it does happen. This is the case here.

We don't want to have huge mocks definition away from a particular test case also away from the actual subtest testing code, this hurts readability. We can't either define mocks from test cases in subtests since this would need a bunch of structs and if blocks in the subtest, making unreadable as well.

The solution is to have just subtests **without** test cases slices or for loop, so everything is locally scoped despite a bit more code duplication.

Example:

```go
t.Run("store trie error", func(t *testing.T) {
	mockStorageStateErr := NewMockStorageState(ctrl)
	mockStorageStateErr.EXPECT().StoreTrie(trieState, &block.Header).Return(errTestDummyError)
	// ...
	s := &Service{storageState: mockStorageStateErr}
	err := s.handleBlock(block, state)
	assert.ErrorIs(t, err, expErr)
	if expErr != nil {
		assert.EqualError(t, err, expErrMsg)
	}
})
```

### Mocks configuration similar from a test case to another

Then configure the mocks in the subtest, and have its expected args / return values in the test cases slice.

File: dot/core/service.go [hunk 3]
```diff
@@ -411,23 +410,24 @@ func (s *Service) maintainTransactionPool(block *types.Block) {
...
+		val, err := rt.ValidateTransaction(tx.Extrinsic)
+		if err != nil {
+			// failed to validate tx, remove it from the pool or queue
+			s.transactionState.RemoveExtrinsic(tx.Extrinsic)
+			continue
+		}
+
+		tx = transaction.NewValidTransaction(tx.Extrinsic, val)
+
+		// Err is only thrown if tx is already in pool, in which case it still gets removed
```
[the remark this answers is not in the document]
Comment (jimjbrettj) [reply dot/core/service.go:428]: This comment already existed, but I believe what was meant is that the only possible error thrown in this case would be if the txn is already in the pool, in which case the behavior is the same

**Comment by jimjbrettj** [comment]
https://github.com/ChainSafe/gossamer/pull/2239#discussion_r803778798

@qdm12 It takes a few seconds, maybe 5-10 on my machine, to generate the wasm. This is why I was skeptical to construct/clean up for every test even though it is cleaner.

I'm actually not sure why we remove them, maybe just due to size? Maybe @noot or @timwu20 could provide more insight on this.

File: dot/core/service_test.go (hunk 2 above)
Comment (jimjbrettj) [reply dot/core/service_test.go]: I like this. I think this would be a good "unit test standard" to help out tests be more consistent across the codebase

File: dot/core/service_test.go [hunk 4]
```diff
@@ -0,0 +1,651 @@
...
+var errTestDummyError = errors.New("test dummy error")
+var testWasmPaths []string
+
+func TestGenerateWasm(t *testing.T) {
+	wasmFilePaths, err := runtime.GenerateRuntimeWasmFile()
+	require.NoError(t, err)
+	testWasmPaths = wasmFilePaths
+}
+
+func Test_Service_StorageRoot(t *testing.T) {
```
Comment (timwu20) [inline dot/core/service_test.go:39]: why is it when I run this test on it's own it fails?
```
go test -timeout 30s -run ^Test_Service_StorageRoot$ github.com/ChainSafe/gossamer/dot/core -v

=== RUN   Test_Service_StorageRoot
=== RUN   Test_Service_StorageRoot/nil_storage_state
=== PAUSE Test_Service_StorageRoot/nil_storage_state
=== RUN   Test_Service_StorageRoot/storage_trie_state_error
=== PAUSE Test_Service_StorageRoot/storage_trie_state_error
=== RUN   Test_Service_StorageRoot/storage_trie_state_ok
=== PAUSE Test_Service_StorageRoot/storage_trie_state_ok
=== CONT  Test_Service_StorageRoot/nil_storage_state
=== CONT  Test_Service_StorageRoot/storage_trie_state_ok
=== CONT  Test_Service_StorageRoot/storage_trie_state_error
=== CONT  Test_Service_StorageRoot/nil_storage_state
    /Users/timothywu/dev/gossamer/dot/core/service.go:148: Unexpected call to *core.MockStorageState.TrieState([<nil>]) at /Users/timothywu/dev/gossamer/dot/core/service.go:148 because:
        expected call at /Users/timothywu/dev/gossamer/dot/core/service_test.go:80 has already been called the max number of times
=== CONT  Test_Service_StorageRoot/storage_trie_state_ok
    /Users/timothywu/dev/gossamer/dot/core/controller.go:137: missing call(s) to *core.MockStorageState.TrieState(is nil) /Users/timothywu/dev/gossamer/dot/core/service_test.go:80
    /Users/timothywu/dev/gossamer/dot/core/controller.go:137: aborting test due to missing call(s)
--- FAIL: Test_Service_StorageRoot (0.00s)
    --- FAIL: Test_Service_StorageRoot/nil_storage_state (0.00s)
    --- PASS: Test_Service_StorageRoot/storage_trie_state_error (0.00s)
    --- FAIL: Test_Service_StorageRoot/storage_trie_state_ok (0.00s)
FAIL
FAIL	github.com/ChainSafe/gossamer/dot/core	0.275s
```

**Review by timwu20 — CHANGES_REQUESTED** [review]

File: dot/core/service_test.go (hunk 4 above)
Comment (jimjbrettj) [reply dot/core/service_test.go:39]: hmm odd, this works for me as well

File: dot/core/service_test.go [hunk 5]
```diff
@@ -0,0 +1,651 @@
...
+			if tt.expErr != nil {
+				assert.EqualError(t, err, tt.expErrMsg)
+			}
+			assert.Equal(t, tt.exp, res)
+		})
+	}
+}
+
+func Test_Service_handleCodeSubstitution(t *testing.T) {
+	testRuntime, err := os.ReadFile(runtime.POLKADOT_RUNTIME_FP)
```
[the remark this answers is not in the document]
Comment (jimjbrettj) [reply dot/core/service_test.go:95]: @timwu20 Could you elaborate on this a bit more, I don't quite understand. Doesn't the Service interface provide a runtime through the config? The only usage of wasmer.NewInstance(code, cfg) I see is in handleCodeSubstitution() which I would think we need since its a change to the runtime. Unless you are saying there is a better way of going about that?

File: dot/core/service_test.go [hunk 6]
```diff
@@ -0,0 +1,651 @@
...
+			exp: common.Hash{0x3, 0x17, 0xa, 0x2e, 0x75, 0x97, 0xb7, 0xb7, 0xe3, 0xd8, 0x4c, 0x5, 0x39, 0x1d, 0x13, 0x9a,
+				0x62, 0xb1, 0x57, 0xe7, 0x87, 0x86, 0xd8, 0xc0, 0x82, 0xf2, 0x9d, 0xcf, 0x4c, 0x11, 0x13, 0x14},
+			retTrieState: ts,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			s := tt.service
+			if tt.name != "nil storage state" {
```
[the remark this answers is not in the document]
Comment (timwu20) [reply dot/core/service_test.go:77]: IMO this is really ugly and difficult to read.  Changing the name of the test would result in test failure? I'd consider not even making this test run in parallel in this case.  The parent `Test_Service_StorageRoot` can run in parallel yes, but I don't see the harm in running the subtests consecutively.

[the remark this answers is not in the document]
Comment (jimjbrettj) [reply dot/core/service_test.go:77]: @timwu20 yup I have changed it to check a parameter like quentin had suggested. Besides making the subtests no longer run in parallel, is there anything else I need to do to clean this up?

File: dot/core/service_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (timwu20) [reply dot/core/service_test.go:95]: create a new `_handleCodeSubstitution` that takes in an extra parameter `func (code []byte, cfg *wasmer.Config) (*wasmer.Instance, error)` and have `handleCodeSubstitution` call that method instead using `wasmer.NewInstance` by default.  In your unit tests you just provide your own `*runtime.Instance`

File: dot/core/service_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jimjbrettj) [reply dot/core/service_test.go:77]: Maybe this will remove flakiness

File: dot/core/service_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jimjbrettj) [reply dot/core/service_test.go:95]: ohhh okay I misunderstood I see. Makes a lot of sense, will implement

File: dot/core/service_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (qdm12) [reply dot/core/service_test.go:77]: I don't see how running tests in parallel or not is relevant here though?

[the remark this answers is not in the document]
Comment (timwu20) [reply dot/core/service_test.go]: I discussed this with @jimjbrettj  over zoom.  With the introduction of gomock controllers and it's reliance on a locally scoped `t *testing.T` we may not be able to use the IDE code generation for tests if we want the subtests to run in parallel.  

Yes detecting race conditions would be better if each subtest is run in parallel, but the tradeoff is we can't use the code generated tests with predefined state if we are using gomock controllers.  In the past I just created my own mock implementations of interfaces that did not rely on `t *testing.T`.  Rather, just emit testing errors when assertions fail or through inspection of the mock implementation.

In this case specifically, it may just make more sense to avoid using the slice of anonymous test structs and just adopt the explicit way of writing out each subtest.  We should avoid repeating code over and over again, which I have discussed, and it will allow us to keep using gomock controllers and run the subtests in parallel.

File: dot/core/service_test.go (hunk 1 above)
Comment (timwu20) [reply dot/core/service_test.go:633]: this should be resolved by removing the reliance on calling `wasmer.NewInstance`.

File: dot/core/service_test.go [hunk 7]
```diff
@@ -0,0 +1,627 @@
...
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+var errTestDummyError = errors.New("test dummy error")
+var testWasmPaths []string
+
+func Test_Service_StorageRoot(t *testing.T) {
```
Comment (timwu20) [inline dot/core/service_test.go:31]: this test can be run in parallel

File: dot/core/service_test.go [hunk 8]
```diff
@@ -0,0 +1,627 @@
...
+			assert.ErrorIs(t, err, tt.expErr)
+			if tt.expErr != nil {
+				assert.EqualError(t, err, tt.expErrMsg)
+			}
+			assert.Equal(t, tt.exp, res)
+		})
+	}
+}
+
+func newTestInstance(code []byte, cfg *wasmer.Config) (*wasmer.Instance, error) {
```
Comment (timwu20) [inline dot/core/service_test.go:88]: nit: move this into the `Test_Service_StorageRoot`

File: dot/core/service_test.go [hunk 9]
```diff
@@ -0,0 +1,627 @@
...
+	return &wasmer.Instance{}, nil
+}
+
+func Test_Service_handleCodeSubstitution(t *testing.T) {
+	testRuntime := []byte{21}
```
Comment (timwu20) [inline dot/core/service_test.go:93]: add `t.Parallel` to the parent as well

File: dot/core/service_test.go [hunk 10]
```diff
@@ -0,0 +1,627 @@
...
+			codeSubstitute:       testCodeSubstitute,
+			blockState:           mockBlockState,
+			codeSubstitutedState: mockCodeSubState,
+		}
+		err := s._handleCodeSubstitution(blockHash, nil, newTestInstance)
+		assert.NoError(t, err)
+	})
+}
+
+func Test_Service_handleBlock(t *testing.T) {
```
Comment (timwu20) [inline dot/core/service_test.go:188]: add `t.Parallel` to the parent as well

File: dot/core/service_test.go [hunk 11]
```diff
@@ -0,0 +1,627 @@
...
+		}
+		err = s.handleBlock(&block, trieState)
+		assert.ErrorIs(t, err, expErr)
+		if expErr != nil {
+			assert.EqualError(t, err, expErrMsg)
+		}
+	})
+}
+
+func Test_Service_HandleBlockProduced(t *testing.T) {
```
Comment (timwu20) [inline dot/core/service_test.go:389]: add `t.Parallel` to the parent as well

File: dot/core/service_test.go (hunk 1 above)
Comment (jimjbrettj) [reply dot/core/service_test.go]: yup yup it is no longer an issue

File: dot/core/service_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (qdm12) [reply dot/core/service_test.go]: > we may not be able to use the IDE code generation for tests if we want the subtests to run in parallel.

It's not necessarily about running in parallel. If one subtest fails because of a mock assertion failing, it might fail the parent test since it would take `t` from the parent test, but would run within the subtest. I guess it's *alriiight* to share a mock controller  at the parent test, although it's not ideal either. For test generation, you could use Github copilot it does help quite a bit to have a good middleground between manual and automated test writing.

> Rather, just emit testing errors when assertions fail or through inspection of the mock implementation.

That's the same problem though, since it would fail the parent test if one of the subtest fails for example. This is the same as passing `t` once to a subtest-local mock controller.

> In this case specifically, it may just make more sense to avoid using the slice of anonymous test structs and just adopt the explicit way of writing out each subtest. 

Agree depending on the case. If mocks configuration is quite different from one test case to another (i.e. a mock depending on another mock, for certain test cases), then yes dumb subtests without for loop / test case slices is best. Otherwise, configure your mock within the subtest. I wrote [this comment](https://github.com/ChainSafe/gossamer/pull/2239#discussion_r803765992) which sums it up from my perspective.

[the remark this answers is not in the document]
Comment (qdm12) [reply dot/core/service_test.go]: Created https://github.com/golang/mock/issues/625 if that interests anyone 😄
But yes, ultimately, controller construction within the subtest is pretty good for local scoping / readability / parallel tests.

File: dot/core/service_test.go [hunk 12]
```diff
@@ -0,0 +1,588 @@
...
+		}
+
+		extrinsic := types.Extrinsic{21}
+		vt := transaction.NewValidTransaction(extrinsic, validity)
+
+		ctrl := gomock.NewController(t)
+		runtimeMock := new(mocksruntime.Instance)
+		runtimeMock.On("ValidateTransaction", types.Extrinsic{21}).Return(nil, errTestDummyError)
+		mockTxnState := NewMockTransactionState(ctrl)
+		mockTxnState.EXPECT().RemoveExtrinsic(types.Extrinsic{21}).MaxTimes(2)
```
Comment (qdm12) [inline dot/core/service_test.go:444]: perhaps?

```suggestion
		mockTxnState.EXPECT().RemoveExtrinsic(types.Extrinsic{21}).Times(2)
```

File: dot/core/service.go [hunk 13]
```diff
@@ -249,6 +248,13 @@ func (s *Service) handleBlock(block *types.Block, state *rtstorage.TrieState) er
 }
 
 func (s *Service) handleCodeSubstitution(hash common.Hash, state *rtstorage.TrieState) error {
+	return s._handleCodeSubstitution(hash, state, wasmer.NewInstance)
```
[the remark this answers is not in the document]
Comment (jimjbrettj) [reply dot/core/service.go:251]: man I just wanted to pretend I was a solidity programmer!

File: dot/core/service.go [hunk 14]
```diff
@@ -249,6 +248,13 @@ func (s *Service) handleBlock(block *types.Block, state *rtstorage.TrieState) er
...
+}
+func (s *Service) _handleCodeSubstitution(
+	hash common.Hash,
+	state *rtstorage.TrieState,
+	newInstance func(code []byte, cfg *wasmer.Config) (*wasmer.Instance, error),
```
[the remark this answers is not in the document]
Comment (jimjbrettj) [reply dot/core/service.go:256]: I'm thinking we should leave as is so that handleCodeSubstitution uses wasmer.NewInstance by default without requiring us to set it manually as a field of Service. If you really think I should change I can, but I think since its only used in the one function and even that's just for testing that its easier to leave as is.

**Comment by jimjbrettj** [comment]
updated!

File: dot/core/service.go [hunk 15]
```diff
@@ -53,7 +55,8 @@ type Service struct {
 	codeSubstitutedState CodeSubstitutedState
 
 	// Keystore
-	keys *keystore.GlobalKeystore
+	keys        *keystore.GlobalKeystore
+	newInstance newWasmerInstanceFunc
```
Comment (timwu20) [inline dot/core/service.go:59]: why did you add this as a `Service` attribute? It's not used anywhere else I believe, so I think it should be localized to `handleCodeSubstitution` as a parameter.

**Review by timwu20 — CHANGES_REQUESTED** [review]

File: dot/core/service.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (timwu20) [reply dot/core/service.go]: I disagree with storing this in `Service`. IMO doesn't make a lot of sense that this is part of the "state" of `Service`.  This is just to make `handleCodeSubsitution` more testable and it should be localised to that function.

File: dot/core/service.go (hunk 15 above)
Comment (timwu20) [reply dot/core/service.go:59]: I think we should change it back. I added a comment in the resolved conversation. https://github.com/ChainSafe/gossamer/pull/2239#discussion_r810232906

File: dot/core/service_test.go (hunk 2 above)
Comment (qdm12) [reply dot/core/service_test.go]: Added in https://github.com/ChainSafe/gossamer/pull/2311

File: dot/core/service_test.go (hunk 4 above)
Comment (qdm12) [reply dot/core/service_test.go:31]: Is it a flaky test still now?

File: dot/core/service_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (qdm12) [reply dot/core/service_test.go]: Added some code style section for mocks configuration in subtests at https://github.com/ChainSafe/gossamer/pull/2311

File: dot/core/service.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (qdm12) [reply dot/core/service.go]: I get your point although having 2 `handleCodeSubstitution` is kind of ugly too.

File: dot/core/service.go [hunk 16]
```diff
@@ -249,6 +252,12 @@ func (s *Service) handleBlock(block *types.Block, state *rtstorage.TrieState) er
 }
 
 func (s *Service) handleCodeSubstitution(hash common.Hash, state *rtstorage.TrieState) error {
+	return s.handleCodeSubstituionWithWasmerInstance(hash, state)
+}
+func (s *Service) handleCodeSubstituionWithWasmerInstance(
```
Comment (qdm12) [inline dot/core/service.go:257]:
```suggestion
func (s *Service) handleCodeSubstitutionWithWasmerInstance(
```

File: dot/core/service.go [hunk 17]
```diff
@@ -249,6 +252,12 @@ func (s *Service) handleBlock(block *types.Block, state *rtstorage.TrieState) er
 }
 
 func (s *Service) handleCodeSubstitution(hash common.Hash, state *rtstorage.TrieState) error {
+	return s.handleCodeSubstituionWithWasmerInstance(hash, state)
```
Comment (qdm12) [inline dot/core/service.go:255]:
```suggestion
	return s.handleCodeSubstitutionWithWasmerInstance(hash, state)
```

**Review by qdm12 — APPROVED** [review]

File: dot/core/service.go [hunk 18]
```diff
@@ -252,11 +250,12 @@ func (s *Service) handleBlock(block *types.Block, state *rtstorage.TrieState) er
 }
 
 func (s *Service) handleCodeSubstitution(hash common.Hash, state *rtstorage.TrieState) error {
-	return s.handleCodeSubstituionWithWasmerInstance(hash, state)
+	return s.handleCodeSubstitutionWithWasmerInstance(hash, state, wasmer.NewInstance)
 }
-func (s *Service) handleCodeSubstituionWithWasmerInstance(
+func (s *Service) handleCodeSubstitutionWithWasmerInstance(
```
Comment (timwu20) [inline dot/core/service.go:255]: can you change it back to `_handleCodeSubstitution`.  It's understood that you're doing this to make the `handleCodeSubstitution` more testable.  The new name implies it's doing something different and is called by other receiver methods.

File: dot/core/service.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (qdm12) [reply dot/core/service.go]: Conclusion:

- pass it as function argument so it's more explicit and localized
- have 2 similarly-named functions which is not too pretty, but worth the point above!

File: dot/core/service.go (hunk 18 above)
Comment (jimjbrettj) [reply dot/core/service.go:255]: will do!

**Review by timwu20 — APPROVED** [review]

File: dot/core/service.go [hunk 19]
```diff
@@ -249,6 +250,13 @@ func (s *Service) handleBlock(block *types.Block, state *rtstorage.TrieState) er
 }
 
 func (s *Service) handleCodeSubstitution(hash common.Hash, state *rtstorage.TrieState) error {
+	return s._handleCodeSubstitution(hash, state, wasmer.NewInstance)
+}
+func (s *Service) _handleCodeSubstitution(
```
Comment (EclesioMeloJunior) [inline dot/core/service.go:255]: I think we don't need `_` for a helper or private functions, you could split the `_handleCodeSubstitution` into `hasCodeSubstitue` and `substitute` then you call them sequentially inside `handleCodeSubstitution` like:

```go
func (s *Service) handleCodeSubstitution ... {
    if hasCodeSubstitue(hash) {
        applySubstitute(hash, state, instance)
    }
}
```

File: dot/core/service_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (qdm12) [reply dot/core/service_test.go]: Tim found a new way to do mocks with a test case slice/map + for loop and subtests. However I don't think it works when a mock needs to return a mock for example.

1. For each mock, add a mock constructor field in your test case with signature `func(ctrl *gomock.Controller) YourInterface`
2. Configure your mock in the test case by defining the function body
3. Call that function from your test case in the subtest, passing it the subtest's Gomock controller as argument

Again, if you need to i.e. return a mock from another mock, this method doesn't seem to work and in this case you need to use one of the two options mentioned in the codestyle document.

File: dot/core/service.go (hunk 19 above)
Comment (jimjbrettj) [reply dot/core/service.go:255]: what do you think @timwu20 @qdm12 ?

File: dot/core/service.go [hunk 20]
```diff
@@ -249,6 +250,13 @@ func (s *Service) handleBlock(block *types.Block, state *rtstorage.TrieState) er
 }
 
 func (s *Service) handleCodeSubstitution(hash common.Hash, state *rtstorage.TrieState) error {
+	return s._handleCodeSubstitution(hash, state, wasmer.NewInstance)
```
Comment (noot) [inline dot/core/service.go:253]: whats the point of this? also `_` prefixed funcs are not really a thing in go afaik

Comment (noot) [reply dot/core/service.go:253]: oh its so you can inject the function during tests i guess?

**Comment by noot** [comment]
the code logic essentially looks fine, i would double check that the extrinsic you're submitting in the stress test `TestSync_SubmitExtrinsic` is the same version and such as in the unit tests here. however it should be a valid transaction, as the `SubmitExtrinsic` function validates it, and will return error if it's invalid, so not entirely sure how your changes would effect that :/ 

also, does `Test_SubmitAndWatchExtrinsic` (other stress test) pass for you on this branch? if so, i think we can skip `TestSync_SubmitExtrinsic`, as `Test_SubmitAndWatchExtrinsic` looks to be an improved version of it

File: dot/core/service.go (hunk 20 above)
Comment (jimjbrettj) [reply dot/core/service.go:253]: yup yup exactly! Makes it so I dont have to generate a valid wasm instance for the mocked unit tests so it goes a lot faster (no more need to generate wasm files)

File: dot/core/service.go (hunk 19 above)
Comment (jimjbrettj) [reply dot/core/service.go]: @EclesioMeloJunior updated!

**Review by EclesioMeloJunior — APPROVED** [review]
